oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2ResourceServerAutoConfig.java

```diff
@@ -45,15 +58,46 @@
     JwtDecoder jwtDecoder(OAuth2ResourceServerProperties oAuth2ResourceServerProperties,
                           OktaOAuth2Properties oktaOAuth2Properties) {
 
-            NimbusJwtDecoderJwkSupport decoder = new NimbusJwtDecoderJwkSupport(oAuth2ResourceServerProperties.getJwt().getJwkSetUri());
-            decoder.setJwtValidator(TokenUtil.jwtValidator(oAuth2ResourceServerProperties.getJwt().getIssuerUri(), oktaOAuth2Properties.getAudience()));
-            decoder.setRestOperations(restOperations());
-            return decoder;
+        NimbusJwtDecoderJwkSupport decoder = new NimbusJwtDecoderJwkSupport(oAuth2ResourceServerProperties.getJwt().getJwkSetUri());
+        decoder.setJwtValidator(TokenUtil.jwtValidator(oAuth2ResourceServerProperties.getJwt().getIssuerUri(), oktaOAuth2Properties.getAudience()));
+        decoder.setRestOperations(restOperations());
+        return decoder;
     }
 
+
     private RestOperations restOperations() {
         RestTemplate restTemplate = new RestTemplate();
         restTemplate.getInterceptors().add(new UserAgentRequestInterceptor());
         return restTemplate;
     }
+
+    @Bean
+    @Conditional(OktaOpaqueTokenIntrospectConditional.class)
+    OpaqueTokenIntrospector opaqueTokenIntrospector(OktaOAuth2Properties oktaOAuth2Properties,
+                                                    OAuth2ResourceServerProperties oAuth2ResourceServerProperties) {
+
+        OpaqueTokenIntrospector delegate = new NimbusOpaqueTokenIntrospector(
+            oAuth2ResourceServerProperties.getOpaquetoken().getIntrospectionUri(),
+            oAuth2ResourceServerProperties.getOpaquetoken().getClientId(),
+            oAuth2ResourceServerProperties.getOpaquetoken().getClientSecret());
+
+        return token -> {
+            OAuth2AuthenticatedPrincipal principal = delegate.introspect(token);
+            Collection<GrantedAuthority> mappedAuthorities =
+                (Collection<GrantedAuthority>) TokenUtil.tokenClaimsToAuthorities(principal.getAttributes(), oktaOAuth2Properties.getGroupsClaim());
+            return new DefaultOAuth2AuthenticatedPrincipal(
+                principal.getName(), principal.getAttributes(), mappedAuthorities);
+        };
+    }
+
+    @Bean
+    RestTemplate restTemplate() {
```

Comment:
Creating a `RestTemplate` bean like this may conflict with other applications.  Where is this one used?  Maybe just create it an instance for that case

Reply:
`RestTemplate` bean type is scanned for by `NimbusOpaqueTokenInstrospector` because of the need to make a network call to remote introspection endpoint. I've put in a conditional on this bean to address your concern/comment.

Reply:
removed it safely (unused).